libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,137 @@
+		return;
+
+	// Find the literal argument of the ``memoryinit`` call, if there is a unique such call, otherwise abort.
+	yulAssert(dynamic_cast<EVMDialect const*>(&_context.dialect), "MemoryEscalator can only be run on EVMDialect objects.");
+	Literal* memoryInitLiteral = nullptr;
+	if (
+		auto memoryInits = FunctionCallFinder::run(*_object.code, "memoryinit"_yulstring);
+		memoryInits.size() == 1
+	)
+		memoryInitLiteral = std::get_if<Literal>(&memoryInits.front()->arguments.back());
+	if (!memoryInitLiteral)
+		return;
+	u256 reservedMemory = valueOfLiteral(*memoryInitLiteral);
+
+	auto callGraph = CallGraphGenerator::callGraph(*_object.code).functionCalls;
+
+	// Collect all names of functions contained in cycles in the callgraph.
+	// TODO: this algorithm is suboptimal and can be improved.
+	std::set<YulString> containedInCycle;
+	auto findCycles = [
+		&,
+		visited = std::map<YulString, uint64_t>{},
+		currentPath = std::vector<YulString>{}
+	](YulString _node, auto& _recurse) mutable -> void
+	{
+		if (auto it = std::find(currentPath.begin(), currentPath.end(), _node); it != currentPath.end())
+			containedInCycle.insert(it, currentPath.end());
+		else
+		{
+			visited[_node] = currentPath.size();
+			currentPath.emplace_back(_node);
+			for (auto const& child: callGraph[_node])
+				_recurse(child, _recurse);
+			currentPath.pop_back();
+		}
+	};
+	findCycles(YulString{}, findCycles);
+
+	using SlotAllocations = map<YulString, map<YulString, u256>>;
+	SlotAllocations slotAllocations;
+	uint64_t requiredSlots = 0;
+
+	// Walks the call graph using a Depth-First-Search.
+	// The visit assigns slots for the variables that need to be moved in the current function.
+	// Each visit resets the number of used slot at the end, resulting in all child visits starting
+	// at the same number of reserved slots.
+	// Since within cycles there are no variables to be moved, this will result in each variable having a unique
+	// offset on each path through the call graph. However, distinct paths may share the same offsets for different
+	// variables.
+	auto allocateMemorySlots = [
+		&,
+		visited = std::set<YulString>{},
+		numUsedSlots = uint64_t(0)
+	](YulString _function, auto& _recurse) mutable -> bool {
+		if (visited.count(_function))
+			return true;
+		visited.insert(_function);
+
+		uint64_t previouslyUsedSlots = numUsedSlots;
+		if (functionStackErrorInfo.count(_function))
+		{
+			if (containedInCycle.count(_function))
+				return false;
+			auto const& stackErrorInfo = functionStackErrorInfo.at(_function);
+			for (auto const& variable: stackErrorInfo.variables)
+				if (variable.empty())
+				{
+					// TODO: Too many function arguments or return parameters.
+				}
+				else
+				{
+					auto& allocations = slotAllocations[_function];
+					if (!allocations.count(variable))
+						allocations[variable] = reservedMemory + 32 * numUsedSlots++;
+				}
+		}
+		requiredSlots = std::max(requiredSlots, numUsedSlots);
```

Comment:
Is there a way to write this procedure without capturing `requiredSlots` and instead passing the old value to the recursive sub-call and returning it from there, computing the max across sub-calls?

Reply:
I can do something like that, yes.